axelrod/strategies/meta.py

```diff
@@ -52,10 +56,48 @@ def __init__(self, team=None):
                     'manipulates_source',
                     'manipulates_state']:
             self.classifier[key] = any(t.classifier[key] for t in self.team)
-
         for t in self.team:
             self.classifier['makes_use_of'].update(t.classifier['makes_use_of'])
 
+    def __eq__(self, other):
```

Comment:
Ok, so it looks like it might just be a problem with the `Qlearners` then:

```python
>>> for s in axl.strategies:
...   for _ in range(3):
...        p1 = s()
...        p2 = s()
...        if p1 != p2:
...            print(p1, p2)
Arrogant QLearner Arrogant QLearner
Arrogant QLearner Arrogant QLearner
Cautious QLearner Cautious QLearner
Hesitant QLearner Hesitant QLearner
Hesitant QLearner Hesitant QLearner
Risky QLearner Risky QLearner
Meta Majority Meta Majority
Meta Majority Meta Majority
Meta Majority Meta Majority
Meta Majority Long Memory Meta Majority Long Memory
Meta Majority Long Memory Meta Majority Long Memory
Meta Majority Long Memory Meta Majority Long Memory
Meta Minority Meta Minority
Meta Minority Meta Minority
Meta Minority Meta Minority
Meta Mixer Meta Mixer
Meta Mixer Meta Mixer
Meta Mixer Meta Mixer
Meta Winner Meta Winner
Meta Winner Meta Winner
Meta Winner Meta Winner
Meta Winner Ensemble Meta Winner Ensemble
Meta Winner Ensemble Meta Winner Ensemble
Meta Winner Ensemble Meta Winner Ensemble
Meta Winner Long Memory Meta Winner Long Memory
Meta Winner Long Memory Meta Winner Long Memory
Meta Winner Stochastic Meta Winner Stochastic
Meta Winner Stochastic Meta Winner Stochastic
Meta Winner Stochastic Meta Winner Stochastic
NMWE Long Memory NMWE Long Memory
NMWE Long Memory NMWE Long Memory
NMWE Long Memory NMWE Long Memory
NMWE Stochastic NMWE Stochastic
NMWE Stochastic NMWE Stochastic
NMWE Stochastic NMWE Stochastic
Nice Meta Winner Nice Meta Winner
Nice Meta Winner Nice Meta Winner
Nice Meta Winner Nice Meta Winner
Nice Meta Winner Ensemble Nice Meta Winner Ensemble
Nice Meta Winner Ensemble Nice Meta Winner Ensemble
Nice Meta Winner Ensemble Nice Meta Winner Ensemble
```
This is arguably a slight bug in the Q learners.

So I suggest you leav the `__eq__`  as I've suggested above but make the following change to the Q learners (I have modified lines 51, 52, 64, 65, 66):

```python
50         self.prev_action = None                                                                                                                 ~_ 51         self.original_prev_action = None                                                                                                        
~   52         self.history = [] # type: List[Action]                                                                                                                                                                                                                        
  ...                                                                                  
   62     def strategy(self, opponent: Player) -> Action:                                                                                                63         """Runs a qlearn algorithm while the tournament is running."""                                                                          
+  64         if len(self.history) == 0:                                                                                                              
+  65             self.prev_action = random_choice()                                                                                                  
+  66             self.original_prev_action = self.prev_action   
```

When running with that change, I get no output from:

```python
>>> for s in axl.strategies:
...   for _ in range(3):
...        p1 = s()
...        p2 = s()
...        if p1 != p2:
...            print(p1, p2)
```

This has the benefit of just having 1 `__eq__` method (which I think should definitely be the case).

As far as I can see you'd need to adjust the tests in `test_match_generator`:

```python
 28     def test_init_with_clone(self):                                                                                                               
 29         tt = axelrod.MatchGenerator(                                                                                                               30             self.players, test_turns, test_game, test_repetitions)                                                                                 31         self.assertEqual(tt.players, self.players)                                                                                                
 32         self.assertEqual(tt.turns, test_turns)                                                                                                    
 33         player = tt.players[0]                                                                                                                    
 34         opponent = tt.opponents[0]                                                                                                                
 35         self.assertEqual(player, opponent)                                                                                                         36         # Check that the two player instances are wholly independent                                                                              
 37         opponent.name = 'Test'                                                                                                                    
 38         self.assertNotEqual(player, opponent)   
```

Also some minor fixes to the tests for Q learners (just checking the initial values).